Rendering of enhanced images

gofal opened the pull request:
Fixes #917
Fixes #742
Fixes #806 
Fixes #1497 

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- The DicomImage held many local fields, which caused issues when rendering several frames in parallel. now all data is processed als class variable and some data is cached in threadsafe dictionaries
- In EnhancedCT, EnhancedMR or BTO some rendering information like windowing is stored in Functional Group Sequences. DicomImage did not read those values, but now they are evaluated. This means, there will not be one render pipeline per DicomImage, but there may be one per frame
- DicomImage hand rendering issues with images compressed with some Jpeg data. Because the PhotometricInterpretation and therefore the renderpipeline was created from the original (compressed) dataset. but the transcoder has to change this value when decoding. But then this updated Photometric Interpretation was not evaluated. Now there is no cached precalucated LUT, but first the frame is decoded and then the render pipeline is created based on the decompressed data.
- Because now the SharedFunctionalGroupSequence and the PerFrameFunctionalGroupSequence are needed, there are now convenience functions to access those values
- DicomImage can cache the decompressed pixel data, the render-LUT or the ready rendered image. The caching strategy can be configured for each of this levels separately

gofal committed:
fix dump to display rendered image

gofal committed:
extract nested class in DicomDump

gofal committed:
use Tuple instead of class

gofal committed:
refactoring

gofal committed:
some rendering benchmarks

gofal committed:
always decode whole pixeldata instead of single frame data in RenderImage

gofal committed:
take other ports for unittests, because port 11112 conflicts with a local dcm4che installation

gofal committed:
- move all the DicomDataset-getter methods into a extension methods class, so that it also can be used for IEnumerable<DicomItem>.
- For backwards compatibility, DicomDataset still has the methods, but they are inlined to the extension method
- a new function in DicomDataset returns the FunctionalGroupSequencesItems

gofal committed:
do not cache pixeldata, but load the renderoptions (windowwidth, windowcenter, rescaleslope, rescaleintercept) from dataset and from functional group for each frame on each RenderImage call

gofal committed:
add missing parameter to make code compilable

gofal committed:
Code cleanup

gofal committed:
ImageManager -> ImageFactory

gofal committed:
store extracted frame pixeldata into dictionary cache

gofal committed:
fix merge problem

gofal committed:
- integrate the GrayscaleRenderOptions into the GenericGrayscalePipeline
- get rid of private fields in DicomImage and store/get data from ConcurrentDictionaries, to make rendering of frames in parallel threadsafe

gofal committed:
add minimal image manipulation feature to dicom-dump application for testing purposes

gofal committed:
optimize dicom-dump tool by using WpfImage instead of ImageSharpImage

gofal committed:
implement CachingMode in DicomImage

gofal committed:
Undo renaming ImageManager to ImageFactory for now.

gofal commented:
Sadly this is a huge PR, but it is a big topic.

I split it up into two parts: **this PR here** is intended to make correct rendering work. 
The **next PR** will then be, optimizing the rendering  by
-  avoiding creating too many buffer-arrays, 
-  caching, 
- reusing shared instances across frames, 
- changing the ImageManager/IImage interface to render into existing images

gofal commented:
Please checkout this code and run your render-tests.

In previous fo-dicom versions you propably will have some workaround-code (like first clone into decompressed image and then render the cloned image, or reading windowing-data from tags and setting properties before calling RenderImage).
But now the rendering should work without any of those workarounds, and even several frames per DicomImage in parallel.

gofal commented:
There is one behavior where I aks your feedback:

because in enhanced images and BTO there are different PerFrame rendering properties, DicomImage now requires different RenderOptions/LUTs per frame, so that the initial rendering of all the frames works properly.

But what, if you have rendered a certain frame, then change the `WindowWidth` property of the DicomImage, what would you expect? Currently only the window property of the last rendered image will be changed. Or would you expect, that the window property on all frames have to be set?

I consider to introduce a property to DicomImage, like  `bool AutoapplyToAllFrames` with which the user can decide if changing the propertiy applies to the current frame or to all frames.

In case, the DicomImage is a "regular" multiframe image, with the rendering options configured in root dataset, there will be the same rendering LUT for all frames. This would be part of the next "Optimization PR" where there will only be one LUT in memory. In that case, seting the property `AutoApplyToAllFrames` to `false` would cause that DicomImage will hold a separate rendering LUT per frame.

gofal commented:
Also, the documentation in the DocFX project will follow as soon as the API is fixed. I am waiting for feedback and guess there will be some changes.

gofal committed:
update changelog

gofal committed:
remove file, crated from VisualStudio as backup

mrbean-bremen commented on .editorconfig:
```diff
@@ -1,8 +1,227 @@
-# top-most EditorConfig file
+# Entfernen Sie die folgende Zeile, wenn Sie EDITORCONFIG-Einstellungen von höheren Verzeichnissen vererben möchten.
```
If you want to leave the comments in, probably translate them into English.

gofal replied:
autogenerated by my visual studio, i'll remove it

mrbean-bremen commented on ChangeLog.md:
```diff
@@ -36,6 +36,10 @@
 * Add better logging for inbound connections (#1561)
 * Added User Identity Negotiation support (#1110)
+* Rendering in DicomImage is threadsafe, several frames can be rendered in parallel (#806)
+* Convenience method for accessing items in Functionsl Group Sequences (#742)
```
typo: Functionsl -> Functional

gofal replied:
thanks

mrbean-bremen commented on FO-DICOM.Core/Imaging/Color32.cs:
```diff
@@ -24,53 +24,29 @@ public Color32(int c)
         /// <summary>Alpha</summary>
         public byte A
```
It might have been better to make the refactoring and styling in an extra PR, to separate functionality changes from stylistic changes - I find it a bit difficult to find the actual changes.

gofal replied:
completely agree. When I noticed how big this PR will be, I stopped doing further styling

mrbean-bremen commented on FO-DICOM.Core/Imaging/ColorSpace.cs:
```diff
@@ -20,25 +20,41 @@ public ColorSpace(string name, params Component[] components)
         public static bool operator ==(ColorSpace a, ColorSpace b)
         {
-            if ((object)a == null || (object)b == null) return false;
+            if ((object)a == null || (object)b == null)
+            {
+                return false;
+            }
+
             return a.Name == b.Name;
```
If already at it:
```cs
return (object)a != null && (object)b != null && a.Name == b.Name;
```

gofal replied:
will be included in next PR

mrbean-bremen commented on FO-DICOM.Core/Imaging/ColorSpace.cs:
```diff
@@ -20,25 +20,41 @@ public ColorSpace(string name, params Component[] components)
         public static bool operator !=(ColorSpace a, ColorSpace b)
         {
-            if ((object)a == null || (object)b == null) return true;
+            if ((object)a == null || (object)b == null)
+            {
+                return true;
+            }
+
             return a.Name != b.Name;
```
Same here:
```cs
return (object)a == null || (object)b == null || a.Name != b.Name;
```

mrbean-bremen commented on FO-DICOM.Core/Imaging/CacheType.cs:
```diff
@@ -0,0 +1,38 @@
+        None = 0,
+
+        /// <summary>
+        /// Specifies the raw uncompressed pixel data (where it needs to be decompressed or read from disk)
```
"Caches" instead of "Specifies"?

gofal replied:
sure!

mrbean-bremen commented on FO-DICOM.Core/Imaging/GrayscaleRenderOptions.cs:
```diff
@@ -116,7 +120,7 @@ public Color32[] ColorMap
         /// </summary>
         /// <param name="dataset">Dataset to extract <see cref="GrayscaleRenderOptions"/> from</param>
```
add frame to doc?

gofal replied:
sure

mrbean-bremen commented on FO-DICOM.Core/Imaging/GrayscaleRenderOptions.cs:
```diff
@@ -203,6 +211,60 @@ public static GrayscaleRenderOptions FromWindowLevel(DicomDataset dataset)
+                options.ModalityLUTSequence = modalityLutSequence;
+            }
+
+            if (dataset.TryGetNonEmptySequence(DicomTag.VOILUTSequence, out DicomSequence voiLutSequence))
```
What if you have a VOI LUT sequence, but no WindowCenter/WindowWidth tags? Wouldn't the VOI LUT be ignored in this case because of the check at the start of the method?
I understand that the logic is the same as in `FromWindowLevel`, so I would ask the same question in that code.
Generally, I don't like it too much that the logic is duplicated here, but maybe it is too expansive to factor out all the tag access to work for both...

gofal replied:
Agree. the whole rendering code is very puzzling, and many useless or repeating conditions. 

the public api here is, that the user can get the GrayScaleRenderOptions explicitly from FunctionGroups, or from MinMax value etc. If the user called this method 'FromFunctionalWindowLevel' but there is no FunctaionGroup, then the whole method returns null.
The man Entry 'FromDataset' does all those checks for certain DicomTags and then calls the first method that fits. Meaning: The method 'FromDataset' checks if there are functional groups and finds some. then calls the Method 'FromFunctaionGroup'. And this method then again checks if there are some Functional Groups.
but this is, because this method is public API and therefore requres some checks.

But in all of those methods the VOILUTSequence is read and stored to 'options' if there is one. But this has to be done in every of this public methods, because each of them can be called directly by a user.

This whole class is one of the examples that I will optimize (in terms of code duplication) after this PR was merged.

When I started this PR I thought that there will only be little changes, and did some little code styling. but then I noticed this will grow big and stopped doing futher thins than just making the logic work correctly.

mrbean-bremen commented on FO-DICOM.Core/Imaging/Render/ImageGraphic.cs:
```diff
@@ -195,11 +189,7 @@ public void Rotate(int angle)
-                if (_rotation >= 360) _rotation -= 360;
-                else if (_rotation < 0) _rotation += 360;
-            }
+            _rotation = _rotation % 360;
```
This certainly is better, though I don't really understand the logic before. If a give an angle of 1 degree, it rotates 90 degrees? I would have expected to either only allow multiples of 90 degrees (or some other measure), or use the nearest rotaion to the angle. Not really related to the review...

gofal replied:
yes, only rotation in 90 degree is allowed, because there is no affine transformation matrix included where all resize, rotate, zoom can be performed.
Only swapping x and y loops and direction is supported.
Agree, rotating by an angle of 1 rotated the whole image.

mrbean-bremen commented on Tests/FO-DICOM.Benchmark/FO-DICOM.Benchmark.csproj:
```diff
@@ -23,9 +27,15 @@
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
+    <None Update="Data\CT0070.dcm">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
```
Why is this needed separately? Same for `multiframe.dcm`.

gofal replied:
autogeneated by visual studio. will remove it

mrbean-bremen commented on Tests/FO-DICOM.Benchmark/OpenFileBenchmarks.cs:
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) 2012-2021 fo-dicom contributors.
+﻿// Copyright (c) 2012-2022 fo-dicom contributors.
```
Is there a reason to change this to 2022 and not to the current year (some below)? If it is needed at all - the file has not been touched.

gofal replied:
i am working on this branch since End of 2022. I updated it to current year. But I stopped doing things like those, when I noticed how big this PR will be. 
I'll revert it for now

mrbean-bremen commented on FO-DICOM.Core/Imaging/GrayscaleRenderOptions.cs:
```diff
@@ -116,7 +120,7 @@ public Color32[] ColorMap
         /// <param name="dataset">Dataset to extract <see cref="GrayscaleRenderOptions"/> from</param>
         /// <returns>New grayscale render options instance</returns>
-        public static GrayscaleRenderOptions FromDataset(DicomDataset dataset)
+        public static GrayscaleRenderOptions FromDataset(DicomDataset dataset, int frame)
```
Wouldn't it make sense to use a default value here? Either 0, or maybe even `null` for the case that you will not try to lookup enhanced multi-frames.

gofal replied:
If you want fo-dicom to find the correct rendering parameters from just any DicomDataset, why should a user want to not to try to lookup for values in functional groups?
If a user wants to get rendering options based on min and max value, then the user can call `FromMinMax(dataset)` directly.
The risk of having a default value is then, that the wrong parameters are returned.

mrbean-bremen replied:
True, and I have no strong opinion here. I just thought it would just behave as before that way.

mrbean-bremen commented on Tools/FO-DICOM.Dump/ImageViewer.xaml:
```diff
@@ -0,0 +1,21 @@
+﻿<UserControl x:Class="FellowOakDicom.Dump.ImageViewer"
```
That will really be helpful!

gofal replied:
😃

mrbean-bremen commented:
> Because now the SharedFunctionalGroupSequence and the PerFrameFunctionalGroupSequence are needed, there are now convenience functions to access those values

I actually think that it could make sense to have such convenience functions (e.g. functions that lookup the tag in the both functional group sequences) in the dataset inestead of only in the Imaging component.

> But what, if you have rendered a certain frame, then change the WindowWidth property of the DicomImage, what would you expect? Currently only the window property of the last rendered image will be changed. Or would you expect, that the window property on all frames have to be set?

I would probably expect it to change in all frames.
In CT images, different VOI LUTs in different slices don't make much sense, as you usually want to compare equal HUs. In (single frame) MR images, I have seen different LUTs to compensate for some artifacts at the end of the image, though if done sensibly, this is done using different modality LUTs instead of different VOI LUTs. Not sure about other enhanced multi-frames - I haven't actually seen them so far.
So, all in all, I consider the case that different VOI LUTs in different slices are actually wanted a fringe case (or at least a less frequently needed case).

> I consider to introduce a property to DicomImage, like bool AutoapplyToAllFrames with which the user can decide if changing the propertiy applies to the current frame or to all frames.

That would of course give the user full control. Even here, I would use the AutoApply option as the default.

mrbean-bremen commented on FO-DICOM.Core/Imaging/DicomImage.cs:
```diff
@@ -52,11 +50,30 @@ public DicomImage(DicomDataset dataset, int frame = 0)
+                else
+                {
+                    // assume, probably incorrectly, that the image is RGB
+                    _pi = PhotometricInterpretation.Rgb;
```
I'm amazed that you are still supporting these old images... I think most software would consider this invalid DICOM.

gofal replied:
It's been there until now, i do not see any drawback if we leave it here, so i just kept it. 
Your vode: should it be removed?

mrbean-bremen commented on FO-DICOM.Core/Imaging/DicomImage.cs:
```diff
@@ -262,87 +268,87 @@ public virtual IImage RenderImage(int frame = 0)
+        /// <param name="frame"></param>
+        private void ResetRenderedImage(int frame)
+        {
+            // if 
```
if ?

gofal replied:
started to write a comment, but then decided it would be better to do a method-comment. 
I tell you that I reviewed my code several times, but i did not notice this single comment.

mrbean-bremen commented on FO-DICOM.Core/Imaging/DicomImage.cs:
```diff
@@ -363,37 +369,39 @@ private static DicomPixelData CreateDicomPixelData(DicomDataset dataset)
-                pixelData.PhotometricInterpretation = PhotometricInterpretation.YbrFull;
-                pixelData.PlanarConfiguration = PlanarConfiguration.Planar;
-            }
+            //// TODO: can this temporary fixes be removed now?
```
Yes, would be nice if this now done by the transcoder...

gofal replied:
As I can see from my tests, the Efferent Decoder does it correctly. 
this was only necessary, because the DicomImage read the Photometric Interpretation from the compressed data, then decompressed the data (where the transcoder updated the Photometric Interpretation) and then applied the wrong logic to the decompressed pixels.
But that should be fixed now.

mrbean-bremen reviewed (commented):
Given that I haven't looked at or used the rendering code before, I may have overlooked any problems, but I'm sure @amoerie will find them :) 
Would be interesting to see the benchmark results before and after the change.

amoerie commented:
I will schedule time to review next week. It's amazing, our custom fork has had support for (per frame) functional group sequences for many years, and I was planning to push this here in a pull request in the next two weeks. (As you may have noticed I've been working on this last week too) What are the chances you do this now 🙂

You beat me to it @gofal !
I'll compare the two implementations then and just discuss the differences. 👍

mrbean-bremen commented:
> It's amazing, our custom fork has had support for (per frame) functional group sequences for many years, and I was planning to push this here in a pull request in the next two weeks.

An obvious lack of communication 😀 

Maybe we can blend the two implementations together. Maybe yours has support for accessing enhanced mulit-frame tag at the dataset level (I would like that), and at least some of the imaging changes are not directly related to that, so maybe there is something to take from both...

amoerie commented:
> > It's amazing, our custom fork has had support for (per frame) functional group sequences for many years, and I was planning to push this here in a pull request in the next two weeks.
> 
> An obvious lack of communication 😀 

Haha I understand why you say this but this is just the first time that I have permission and time to push this work upstream. So no bad intentions or anything. 🙂

> 
> Maybe we can blend the two implementations together. Maybe yours has support for accessing enhanced mulit-frame tag at the dataset level (I would like that), and at least some of the imaging changes are not directly related to that, so maybe there is something to take from both...

I'll have to make a detailed comparison and then I'll post my findings here 🙂 I'm also not really an imaging or window leveling expert, but I can talk to people who are if need be.

gofal commented:
> I actually think that it could make sense to have such convenience functions (e.g. functions that lookup the tag in the both functional group sequences) in the dataset inestead of only in the Imaging component.

this convenience function is part of `DicomDataset` and therefore can be used everywhere where you are accessing data in Enhanced datasets.
E.g. if you are doing measurement, you fill often find the pixelspacing stored in functional groups etc.

I am sure, the implementation of amoerie will be more convenient, since the method here is only the minimal necessary to iterate all the values contained in functional groups.

gofal committed:
address code review comments

amoerie commented on FO-DICOM.Core/DicomDataset.cs:
```diff
@@ -177,11 +177,8 @@ public bool AutoValidate
-            return _items.TryGetValue(tag, out DicomItem dummyItem) ? dummyItem as T : null;
-        }
+        public T GetDicomItem<T>(DicomTag tag) where T : DicomItem
+            => DicomDatasetGetterExtensions.GetDicomItem<T>(this, tag);
```
Oompf I'm not a fan of this at all.
I suppose I'll found out further in the pull request why this was necessary...

amoerie commented on FO-DICOM.Core/DicomDataset.cs:
```diff
@@ -1506,6 +1152,613 @@ internal void OnBeforeSerializing()
     }
 
 
+    public static class DicomDatasetGetterExtensions
```
I'm sorry but I think this is a step back. Conceptually these methods belong to DicomDataset, everything in fo-dicom is built around this core container class. 

Technically, the Items in Dicomdataset is a dictionary. The values of a dictionary returns an instance of `ICollection`.
While all of these methods surely work with IEnumerable, we have just lost the assumption that we can safely call Count without enumerating, etc.
Secondly, it is now no longer possible to do any meaningful caching or optimization that requires a field or property on Dicomdataset.

I believe that extensions have a place, for types that you don't own or where the "extension method pattern" makes a lot of sense. I don't think any of these 2 criteria apply here.

gofal replied:
The DicomDataset is a class representing a DICOM Dataset, exactly what it's name says. It is not some BaseClass, where each other class has to inerhit from (for some reason). it is very common, because in DICOM-Standard there are many such Datasets and therefore this class is very often used.

Yes, a DicomDataset is a wrapper around a dictionary and with plenty of additional properties or fields. As we know, building up dictionaries take some time (hashing) and consume some memory (the hash-table is a huge array in case of sparse keys)
If you have the use-case that you have some list of DicomItens and you want to get a value from that list of DicomItems, then it is a lot of overhead to transform that list of Items into a Dictionary (allocating hash-table etc) and the DicomDataset consumes more memory than just an Enumerator of DicomItem - and as we know you love to reduce memory consumption.

So when you want to look for an item contained in one of the function groups, then the most efficient way (instead of building instances of DicomDatasets and copying the whole stuff into a dictionay) would be to iteate through the values. This is why the convenience function instanciates an Enumerator and yields through the items.
So when you then want to get a certain value ot of that IEnumerable, you need the various Get*-Methods. 
There we could either duplicate code (not so nice), or extract these Get*-Methods into some form, where they can be use for DicomDatasets as well as for any IEnumerable<DicomItem>.

I am unsure, which type of Optimization or Caching you mean, that is no longer possible. 
If there were *only* the extension methods, then sure. But in order not to break the API of DicomDataset, and to still have possiblility to add custom code there, there are still those methods, that at the moment just call the extension method.
But in case something would be needed here, we could add some caching or optimization there.

So the extension method apply here, because the getter methods should be applyable also on an IEnumerator not only on a full DicomDataset, and we do not own the IEnumerator.

amoerie replied:
OK I hear your arguments and can somewhat agree that these "get" methods can be extension methods, with the benefit of being able to use them on any list of DicomItem so you don't need to introduce the overhead of a hash table/dictionary.

I still have some questions/remarks though:

1) ~~Can you point me to an example though where one of these extensions is used and the value is NOT a Dicomdataset? Because as far as I can see the functional values method is always used on an actual dataset. What triggered these changes originally?~~ Never mind I missed where the IEnumerable of DICOM items is used in GrayScaleRenderOptions. I'm unsure if this was the correct solution to this problem but at least there is a use case. :-) 

2) Are we dealing with multiple enumeration inside all these methods? Every time you call a method on an `IEnumerable`, it starts up a new enumerator and it has to loop through the items again. `IDictionary` or `ICollection` have their contents buffered internally so it makes no difference for these implementations. JetBrains Rider is giving loads of warnings like this: 

![image](https://github.com/fo-dicom/fo-dicom/assets/2063584/5ae11257-02c8-4b45-a001-e8eec392010c)

If we want to assume that the contents are at least buffered, then maybe these extensions should not be made for IEnumerable but for ICollection?

3) Don't we lose a lot of performance by not using the internal dictionary of Dicomdataset anymore? For example, TryGetValue now does this:

```
        private static bool TryGetValue(this IEnumerable<DicomItem> items, DicomTag key, out DicomItem value)
        {
            value = items.FirstOrDefault(i => i.Tag == key);
            return value != null;
        }
```

To be honest I am flabbergasted that we maintain a dictionary of elements by tag, and then we don't use that dictionary for lookups, or were you still intending to change this?

> I am unsure, which type of Optimization or Caching you mean, that is no longer possible.
If there were only the extension methods, then sure. But in order not to break the API of DicomDataset, and to still have possiblility to add custom code there, there are still those methods, that at the moment just call the extension method.
But in case something would be needed here, we could add some caching or optimization there.

Yes I don't have a specific example, but I was thinking along the lines of being able to cache the functional values or something in a field, and then reusing that field instead of always looping over the items. You can ignore this remark, it was purely hypothetical and like you say, the instance methods still exist.

amoerie commented on FO-DICOM.Core/DicomDatasetExtensions.cs:
```diff
@@ -32,7 +32,7 @@ public static DicomDataset Clone(this DicomDataset dataset)
         /// <param name="time">Tag associated with time value.</param>
         /// <returns>Composite <see cref="System.DateTime"/>.</returns>
-        public static DateTime GetDateTime(this DicomDataset dataset, DicomTag date, DicomTag time)
+        public static DateTime GetDateTime(this IEnumerable<DicomItem> dataset, DicomTag date, DicomTag time)
```
If you do this, then file + class can no longer be named DicomDatasetextensions

gofal replied:
Agree, i am open to rename it.

amoerie commented on FO-DICOM.Core/Imaging/GrayscaleRenderOptions.cs:
```diff
@@ -125,6 +130,11 @@ public static GrayscaleRenderOptions FromDataset(DicomDataset dataset)
                 // If dataset contains WindowWidth and WindowCenter valid attributes used initially for the grayscale options
                 grayscaleRenderOptions = FromWindowLevel(dataset);
             }
+            else if (dataset.FunctionalGroupValues(frame).TryGetValue(DicomTag.WindowWidth, 0, out double functionalWindowWidth) && functionalWindowWidth >= 1.0
```
You're fetching the functional group values twice here, once for the width and once for the height

gofal replied:
It's an enumerator. This is comparable to do a for-loop twice. I thought about calling the FunctionGroupValues(frame) method once, turn it into a List<DicomItem> and then accessing this list.
But this also causes some overhead (instanciating and building list). I thought, for just 2 accesses this whould be more efficent. 
But I can give it a try with Benchmark

amoerie commented on FO-DICOM.Core/Imaging/LUT/VOISequenceLUT.cs:
```diff
@@ -76,7 +76,7 @@ public void Recalculate()
             GetLUTDescriptor();
 
-            var LUTDataElement = _renderOptions.VOILUTSequence.Items[0].GetDicomItem<DicomElement>(DicomTag.LUTData);
+            var LUTDataElement = _voiLUTItem.GetDicomItem<DicomElement>(DicomTag.LUTData);
```
nitpick: variables should begin with lowercase

gofal replied:
agree. "LUT" is an abbreviation usualy written in upper letters. so used to it. Will rename to "lutDataElement".

amoerie commented on FO-DICOM.Core/DicomDataset.cs:
```diff
@@ -633,75 +339,69 @@ public string GetString(DicomTag tag)
         {
-            if (tag.IsPrivate)
+            // gets all items from SharedfunctionalGroup 
+            if (this.TryGetSequence(DicomTag.SharedFunctionalGroupsSequence, out var sharedFunctionalGroupsSequence))
```
Shouldn't `PerFrameFunctionalGroupsSequence` have a higher priority?

(This is based on our fork, I haven't checked the standard)

gofal replied:
sure, checking the standard first is always a good idea ;-)

https://dicom.nema.org/medical/dicom/2020b/output/chtml/part03/sect_C.7.6.16.html
> A single Functional Group shall not be included in both the Shared Functional Groups Sequence (5200,9229) and the Per-frame Functional Groups Sequence (5200,9230).

So the items are either in Shared or in PerFrame, but never in both (and if, then this in invalid and there is no "this has more priority than that). That means, the order of the FunctionalGroupSequence does not matter.

amoerie commented on FO-DICOM.Core/Imaging/GrayscaleRenderOptions.cs:
```diff
@@ -203,6 +212,60 @@ public static GrayscaleRenderOptions FromWindowLevel(DicomDataset dataset)
+            var bits = BitDepth.FromDataset(dataset);
+            var options = new GrayscaleRenderOptions(bits)
+            {
+                RescaleSlope = dataset.Contains(DicomTag.RescaleSlope)
```
The rescale slope and rescale intercept can also be in the first `PixelValueTransformationSequence` item, and take higher priority if those are present.

(This is based on our fork, I haven't checked the standard)

gofal replied:
As I understand the DICOM-standard, then in case of multiframe images there can be attributes in the Functional Groups, either in the Shared or in the PerFrame. Or the value is in Root-Dataset without the FunctionGroups.
Please point me to where you can find something like priority. That the one have has priority over the other. I understand it as duplicate values (in root Dataset AND in the Functional Groups) are invalid with undefined behavior.

amoerie commented on FO-DICOM.Core/DicomDataset.cs:
```diff
@@ -633,75 +339,69 @@ public string GetString(DicomTag tag)
+                && perFrameFunctionalGroupsSequence.Count() > frame)
             {
-                throw new DicomDataException($"Tag: {tag} not found in dataset");
+                var frameFunctionGroupItem = perFrameFunctionalGroupsSequence.ElementAt(frame);
```
Use `perFrameFunctionalGroupsSequence.Items[frame]` to avoid the unnecessary enumerator and for faster accessing.

gofal replied:
agree

amoerie commented on FO-DICOM.Core/DicomDataset.cs:
```diff
@@ -633,75 +339,69 @@ public string GetString(DicomTag tag)
-        {
-            if (!_items.TryGetValue(tag, out item))
+            if (this.TryGetSequence(DicomTag.PerFrameFunctionalGroupsSequence, out var perFrameFunctionalGroupsSequence)
+                && perFrameFunctionalGroupsSequence.Count() > frame)
```
Use `perFrameFunctionalGroupsSequence.Items.Count`

gofal replied:
agree

amoerie commented on FO-DICOM.Core/DicomDataset.cs:
```diff
@@ -633,75 +339,69 @@ public string GetString(DicomTag tag)
             {
-                var privateTag = GetPrivateTag(tag, false);
-                if (privateTag == null)
+                var sharedFunctionGroupItem = sharedFunctionalGroupsSequence.First();
```
Use `sharedFunctionalGroupsSequence.Items[0] ?? throw new ChooseAGoodException("Empty sequence")`

gofal replied:
agree

amoerie commented on FO-DICOM.Core/Imaging/GrayscaleRenderOptions.cs:
```diff
@@ -152,14 +162,11 @@ public static GrayscaleRenderOptions FromDataset(DicomDataset dataset)
-            if (grayscaleRenderOptions.ModalityLUTSequence != null && dataset.TryGetSingleValue(DicomTag.SOPClassUID, out DicomUID sopClassUID))
+            if (grayscaleRenderOptions.ModalityLUTSequence != null
+                && dataset.TryGetSingleValue(DicomTag.SOPClassUID, out DicomUID sopClassUID) 
+                && sopClassUID.UID.IsOneOf(DicomUID.XRayAngiographicImageStorage.UID, DicomUID.XRayRadiofluoroscopicImageStorage.UID, DicomUID.XRayAngiographicBiPlaneImageStorageRETIRED.UID))
```
This now always allocates an array with 3 DICOM UIDs, enumerating them one by one in a loop.
If the triple equality check bothered you, a static readonly HashSet would be a better fit imho.

gofal replied:
agree. The readablility of the code was very very important when doing this rework. This is important to keep the overview.
As mentioned in the introduction, the one single main goal if this PR is, to make the corrent rendering work. 
If I also included optimizations too much here, then this PR would be much bigger and harder to review/understand
So when the logic here is fine, i build up my local rendering test and do opmimization with benchmark and without regression in the rendering tests.
But first I need to know, if the rendering still has some bugs.

amoerie commented on FO-DICOM.Core/DicomAnonymizer.cs:
```diff
@@ -382,8 +383,8 @@ protected static void BlankItem(DicomDataset dataset, DicomItem item, bool nonZe
             var t = item.GetType();
-            return t == typeof(DicomOtherByte) || t == typeof(DicomOtherDouble) || t == typeof(DicomOtherFloat)
-                   || t == typeof(DicomOtherLong) || t == typeof(DicomOtherWord) || t == typeof(DicomUnknown);
+            return t.IsOneOf(typeof(DicomOtherByte), typeof(DicomOtherDouble), typeof(DicomOtherFloat), 
```
This code now always allocates an array of 6 types and loops through it to do the check.
Before, the compiler was free to optimize this in any way it saw possible.
If you still want to get rid of the 6 equality checks, a static readonly HashSet seems like the best option to me.

gofal replied:
same here. Optimization was no priority in this PR, but correctness, and readability. this is a static method, I am sure, the 6 types array is instanciated only once.

amoerie replied:
> I am sure, the 6 types array is instanciated only once.

You seemed so confident saying this that you made me doubt myself hahaha, so obviously I had to check this. :-) 

I wrote a little benchmark that implements IsOtherElement with 3 strategies:
1) Using simple equality checks (aka the original code)
2) Using a static readonly HashSet
3) Using the IsOneOf LINQ extension

This is the full benchmark code:

```csharp
[MemoryDiagnoser]
public class IsOtherElementBenchmarks
{
    private static readonly DicomItem[] DicomItems = {
        new DicomOtherByte(DicomTag.SelectorOBValue, 0),
        new DicomOtherDouble(DicomTag.SelectorODValue, 1.0),
        new DicomOtherFloat(DicomTag.SelectorOFValue, 1.0f),
        new DicomOtherLong(DicomTag.SelectorOLValue, 1),
        new DicomOtherWord(DicomTag.SelectorOWValue, 1),
        new DicomUnknown(DicomTag.SelectorUNValue, 1),
        new DicomCodeString(DicomTag.SelectorCSValue, "123"),
    };

    [Benchmark]
    public bool[] Equality()
    {
        var results = new bool[DicomItems.Length];
        for (int index = 0; index < DicomItems.Length; index++)
        {
            DicomItem item = DicomItems[index];
            results[index] = IsOtherElementEquality(item);
        }
        return results;
    }

    [Benchmark]
    public bool[] HashSet()
    {
        var results = new bool[DicomItems.Length];
        for (int index = 0; index < DicomItems.Length; index++)
        {
            DicomItem item = DicomItems[index];
            results[index] = IsOtherElementHashSet(item);
        }
        return results;
    }

    [Benchmark]
    public bool[] IsOneOf()
    {
        var results = new bool[DicomItems.Length];
        for (int index = 0; index < DicomItems.Length; index++)
        {
            DicomItem item = DicomItems[index];
            results[index] = IsOtherElementIsOneOf(item);
        }
        return results;
    }

    private static bool IsOtherElementEquality(DicomItem item)
    {
        var t = item.GetType();
        return t == typeof(DicomOtherByte) || t == typeof(DicomOtherDouble) || t == typeof(DicomOtherFloat)
               || t == typeof(DicomOtherLong) || t == typeof(DicomOtherWord) || t == typeof(DicomUnknown);
    }

    private static readonly HashSet<Type> OtherElementTypes = new HashSet<Type>(new[]
    {
        typeof(DicomOtherByte), typeof(DicomOtherDouble), typeof(DicomOtherFloat),
        typeof(DicomOtherLong), typeof(DicomOtherWord), typeof(DicomUnknown)
    });

    private static bool IsOtherElementHashSet(DicomItem item)
    {
        var t = item.GetType();
        return OtherElementTypes.Contains(t);
    }

    private static bool IsOtherElementIsOneOf(DicomItem item)
    {
        var t = item.GetType();
        return IsOneOf(t, typeof(DicomOtherByte), typeof(DicomOtherDouble), typeof(DicomOtherFloat),
            typeof(DicomOtherLong), typeof(DicomOtherWord), typeof(DicomUnknown));
    }

    private static bool IsOneOf<T>(T value, params T[] values)
        => values.Contains(value);
}
```

And these are the results: 

``` ini

BenchmarkDotNet=v0.13.5, OS=Windows 11 (10.0.22621.1702/22H2/2022Update/SunValley2)
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET SDK=7.0.302
  [Host]     : .NET 7.0.5 (7.0.523.17405), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.5 (7.0.523.17405), X64 RyuJIT AVX2


```
|   Method |      Mean |    Error |   StdDev |   Gen0 | Allocated |
|--------- |----------:|---------:|---------:|-------:|----------:|
| Equality |  54.39 ns | 0.909 ns | 0.851 ns | 0.0051 |      32 B |
|  HashSet |  96.05 ns | 0.779 ns | 0.729 ns | 0.0050 |      32 B |
|  IsOneOf | 270.99 ns | 3.100 ns | 2.748 ns | 0.0854 |     536 B |

Unfortunately it looks like it does always allocate the array every time! :-( How nice it would be if the compiler could optimize this.

Disclaimer: Yes, I know this method executes within nano seconds so in the grand scheme of things it probably doesn't matter a lot. But in this specific case, you made it multiple times less efficient while I don't see a lot of of benefit compared to the original code.

amoerie commented on FO-DICOM.Core/DicomDataset.cs:
```diff
@@ -1506,6 +1152,613 @@ internal void OnBeforeSerializing()
+        /// Gets the <see cref="DicomItem"/> of the specified <paramref name="tag"/>.
+        /// </summary>
+        /// <typeparam name="T">Type of the return value. Must inherit from <see cref="DicomItem"/>.</typeparam>
+        /// <param name="tag">Requested DICOM tag.</param>
```
Nitpick: all of these extensions methods are missing a `<param>` entry for the `this` parameter it seems.

gofal replied:
agree. A LOT of documentation is missing.
will add the param-entries

amoerie commented on FO-DICOM.Core/DicomDatasetExtensions.cs:
```diff
@@ -53,7 +53,7 @@ public static DateTime GetDateTime(this DicomDataset dataset, DicomTag date, Dic
         /// <param name="topLevelDataset">The top-level dataset (if different from the current dataset). This is where the time zone information will be located</param>
         /// <returns>Composite <see cref="System.DateTimeOffset"/>.</returns>
-        public static DateTimeOffset GetDateTimeOffset(this DicomDataset dataset, DicomTag date, DicomTag time, DicomDataset topLevelDataset = null)
+        public static DateTimeOffset GetDateTimeOffset(this IEnumerable<DicomItem> dataset, DicomTag date, DicomTag time, DicomDataset topLevelDataset = null)
```
Same note here with the missing `<param>` in every method

gofal replied:
agree. this was missing for a while already, but will add it with this PR now.

amoerie reviewed (changes requested):
I took some time to review this, but I still have some things left to compare.
For example, we go about totally differently with VoiLUTSequences. 

In any case, most of it looks good, except for maybe the extensions methods. I may need some time to get used to it, but my initial reaction is ... not so enthusiastic :-)

gofal commented:
@amoerie 
>In any case, most of it looks good, except for maybe the extensions methods. I may need some time to get used to it, but my initial reaction is ... not so enthusiastic :-)

This PR is primarily about image rendering. With multiframe images, the need to access Functional Groups in a convenient and efficient way is high. This is why I added a minimal efficient method. I explained it above.
When you mentioned that you also had implemented some methods for accessing Functional Groups, I posted that I am sure, your implementation would be more convenient.
Maybe I canged you mind when I explained you the reasons for this implementation? Otherwise I am keen so see you implementation than we can discuss which one is better - or we find some merged solution with the best of the two approaches.

gofal commented:
> For example, we go about totally differently with VoiLUTSequences.

PLEASE go here more into detail. Do you have some sample images, where the rendering of hte VoiLutSequences is wrong? I really would appreciate to include them into my rendering-tets and to fix the logic.

As mentioned several times: the aim of this PR is to make the rendering work correctly. And for that I would be very thankfull if you would check this code against your images and report some wrong rendered images.

amoerie commented:
> @amoerie
> 
> > In any case, most of it looks good, except for maybe the extensions methods. I may need some time to get used to it, but my initial reaction is ... not so enthusiastic :-)
> 
> This PR is primarily about image rendering. With multiframe images, the need to access Functional Groups in a convenient and efficient way is high. This is why I added a minimal efficient method. I explained it above. When you mentioned that you also had implemented some methods for accessing Functional Groups, I posted that I am sure, your implementation would be more convenient. Maybe I canged you mind when I explained you the reasons for this implementation? Otherwise I am keen so see you implementation than we can discuss which one is better - or we find some merged solution with the best of the two approaches.

Sure, here is our implementation that parses gray scale render options from the functional group sequences. 
I'm still not sure why you needed the extension methods, because inside a DicomSequence, every item is another Dicomdataset. So we still have all the conventional utility methods for extracting information.

This is GrayScaleRenderOptions.FromDataset:

(Note how we do something totally different with the VOILUTSequence, but I'll address that in a separate comment) 

```csharp
/// <summary>
/// Create <see cref="GrayscaleRenderOptions"/>  from <paramref name="dataset"/> and populate the options properties with values:
/// Bit Depth
/// Rescale Slope
/// Rescale Intercept
/// Window Width
/// Window Center
/// </summary>
/// <param name="dataset">Dataset to extract <see cref="GrayscaleRenderOptions"/> from</param>
/// <returns>New grayscale render options instance</returns>
public static GrayscaleRenderOptions FromDataset(DicomDataset dataset, int frame = 0)
{
    var grayscaleRenderOptions = FromDatasetInternal();

    /*
        David Clunie in comp.protocols.dicom (2000-12-13)
        https://groups.google.com/g/comp.protocols.dicom/c/UBxhOZ2anJ0/m/D0R_QP8V2wIJ
        --------------------------------------------------

        Modality LUTs in XA and XRF objects are totally screwy and do not follow the normal rules.
        [...]
        A Modality LUT may be included with the image to allow it to be scaled back to its proportional value to X-Ray beam intensity.
        In other words, for the objects that use this module (XA and XRF), the Modality LUT is used BACKWARDS.
        It is used to convert stored pixels to X-Ray beam intensity space, but it is NOT APPLIED to stored pixels for the purpose of display
        (or more specifically prior to application of the VOI LUT Module attributes to the stored pixel data).
    */
    if (grayscaleRenderOptions.ModalityLUTSequence != null && dataset.TryGetSingleValue(DicomTag.SOPClassUID, out DicomUID sopClassUID))
    {
        if (sopClassUID.UID == DicomUID.XRayAngiographicImageStorage.UID
            || sopClassUID.UID == DicomUID.XRayRadiofluoroscopicImageStorage.UID
            || sopClassUID.UID == DicomUID.XRayAngiographicBiPlaneImageStorageRETIRED.UID)
        {
            grayscaleRenderOptions.ModalityLUTSequence = null;
        }
    }

    return grayscaleRenderOptions;

    GrayscaleRenderOptions FromDatasetInternal()
    {
        if (dataset.Contains(DicomTag.VOILUTSequence))
        {
            var grayscaleRenderOptionsByVoiLutSequence = FromVoiLutSequence(dataset);
            if (grayscaleRenderOptionsByVoiLutSequence != null)
                return grayscaleRenderOptionsByVoiLutSequence;
        }

        if (dataset.Contains(DicomTag.WindowWidth) && dataset.Contains(DicomTag.WindowCenter))
        {
            var grayscaleRenderOptionsByWindowLevel = FromWindowLevel(dataset);
            if (grayscaleRenderOptionsByWindowLevel != null)
                return grayscaleRenderOptionsByWindowLevel;
        }
        else
        {
            if (dataset.Contains(DicomTag.PerFrameFunctionalGroupsSequence))
            {
                //Check if there are per frame defined WL values
                var perFrameFunctionalSq =
                    dataset.GetDicomItem<DicomSequence>(DicomTag.PerFrameFunctionalGroupsSequence);
                if (perFrameFunctionalSq.Items.Count >= frame)
                {
                    if (TryGetGrayscaleRenderOptionsFromFunctionalGroupSequence(
                            dataset,
                            perFrameFunctionalSq.Items[frame],
                            out var fromWindowLevel
                        ))
                        return fromWindowLevel;
                }
            }

            if (dataset.Contains(DicomTag.SharedFunctionalGroupsSequence))
            {
                //Check if there are shared frame defined WL values
                var sharedFunctionalSq =
                    dataset.GetDicomItem<DicomSequence>(DicomTag.SharedFunctionalGroupsSequence);
                if (sharedFunctionalSq.Items.Count > 0 &&
                    TryGetGrayscaleRenderOptionsFromFunctionalGroupSequence(
                        dataset,
                        sharedFunctionalSq.Items[0],
                        out var fromWindowLevel
                    ))
                    return fromWindowLevel;
            }
        }

        if (dataset.TryGetSingleValue(DicomTag.SmallestImagePixelValue, out int smallest) &&
            dataset.TryGetSingleValue(DicomTag.LargestImagePixelValue, out int largest)
            && smallest < largest)
        {
            //If dataset contains valid SmallestImagePixelValue and LargestImagePixelValue attributes, use range to calculate
            //WindowWidth and WindowCenter
            return FromImagePixelValueTags(dataset);
        }

        //If reached here, minimum and maximum pixel values calculated from pixels data to calculate
        //WindowWidth and WindowCenter
        return FromMinMax(dataset);
    }
}
``` 

And this is the method TryGetGrayscaleRenderOptionsFromFunctionalGroupSequence:

```csharp
private static bool TryGetGrayscaleRenderOptionsFromFunctionalGroupSequence(DicomDataset fullDataSet, DicomDataset functionalGroupDataSet, out GrayscaleRenderOptions fromWindowLevel)
{
    fromWindowLevel = null;

    if (functionalGroupDataSet.Contains(DicomTag.FrameVOILUTSequence))
    {
        var frameVoiLutSq = functionalGroupDataSet.GetDicomItem<DicomSequence>(DicomTag.FrameVOILUTSequence);
        if (frameVoiLutSq.Items.Count > 0)
        {
            var frameVoiLutSqItem = frameVoiLutSq.Items[0];
            if (frameVoiLutSqItem.Contains(DicomTag.WindowWidth) && frameVoiLutSqItem.Contains(DicomTag.WindowCenter))
            {
                var windowWidth = frameVoiLutSqItem.GetValue<double>(DicomTag.WindowWidth, 0);
                var windowCenter = frameVoiLutSqItem.GetValue<double>(DicomTag.WindowCenter, 0);
                var voiLutFunction = frameVoiLutSqItem.GetValueOrDefault<string>(DicomTag.VOILUTFunction, 0, null);

                double? rescaleSlope = null;
                double? rescaleIntercept = null;
                //Check for Rescale Slope/Intercept inside the PixelValueTransformationSequence
                if (functionalGroupDataSet.Contains(DicomTag.PixelValueTransformationSequence))
                {
                    var framePixelTransformationSq =
                        functionalGroupDataSet.GetDicomItem<DicomSequence>(DicomTag.PixelValueTransformationSequence);
                    if (framePixelTransformationSq.Items.Count > 0)
                    {
                        var framePixelTransformationSqItem = framePixelTransformationSq.Items[0];
                        rescaleSlope = framePixelTransformationSqItem.GetValueOrDefault<double?>(DicomTag.RescaleSlope, 0, null);
                        rescaleIntercept = framePixelTransformationSqItem.GetValueOrDefault<double?>(DicomTag.RescaleIntercept, 0, null);
                    }
                }

                var bits = BitDepth.FromDataset(fullDataSet);
                fromWindowLevel = new GrayscaleRenderOptions(bits)
                {
                    RescaleSlope = rescaleSlope ?? fullDataSet.GetSingleValueOrDefault(DicomTag.RescaleSlope, 1.0),
                    RescaleIntercept = rescaleIntercept ?? fullDataSet.GetSingleValueOrDefault(DicomTag.RescaleIntercept, 0.0),

                    WindowWidth = windowWidth,
                    WindowCenter = windowCenter,

                    VOILUTFunction = voiLutFunction ?? fullDataSet.GetSingleValueOrDefault(DicomTag.VOILUTFunction, "LINEAR"),
                    ColorMap = GetColorMap(fullDataSet)
                };

                if (fromWindowLevel.RescaleSlope == 0.0)
                    fromWindowLevel.RescaleSlope = 1.0;

                if (fullDataSet.TryGetSequence(DicomTag.ModalityLUTSequence, out DicomSequence modalityLutSequence))
                    fromWindowLevel.ModalityLUTSequence = modalityLutSequence;
                return true;
            }
        }
    }

    return false;
}
```

I'd be interested to hear your thoughts on this implementation. This has been running in production for a few years already.

amoerie commented:
> > For example, we go about totally differently with VoiLUTSequences.
> 
> PLEASE go here more into detail. Do you have some sample images, where the rendering of hte VoiLutSequences is wrong? I really would appreciate to include them into my rendering-tets and to fix the logic.
> 
> As mentioned several times: the aim of this PR is to make the rendering work correctly. And for that I would be very thankfull if you would check this code against your images and report some wrong rendered images.

As you can imagine, this requires some spelunking through our internal issue tracker and trying to find some old images that were broken before we made changes. I'm not intentionally hiding information from you, it just takes time to collect the test data from years ago. I also still want to talk to the person who made the actual code changes back then.

But you can already see in my code example above how we construct GrayScaleRenderOptions from the VOILUTSequence if it is present, and it takes precedence over all other DICOM tags. 
I'm fairly confident we have other changes in other places too, but I'm still puzzling that out. I'll keep you posted.

amoerie commented:
> > > For example, we go about totally differently with VoiLUTSequences.
> > 
> > 
> > PLEASE go here more into detail. Do you have some sample images, where the rendering of hte VoiLutSequences is wrong? I really would appreciate to include them into my rendering-tets and to fix the logic.
> > As mentioned several times: the aim of this PR is to make the rendering work correctly. And for that I would be very thankfull if you would check this code against your images and report some wrong rendered images.
> 
> As you can imagine, this requires some spelunking through our internal issue tracker and trying to find some old images that were broken before we made changes. I'm not intentionally hiding information from you, it just takes time to collect the test data from years ago. I also still want to talk to the person who made the actual code changes back then.
> 
> But you can already see in my code example above how we construct GrayScaleRenderOptions from the VOILUTSequence if it is present, and it takes precedence over all other DICOM tags. I'm fairly confident we have other changes in other places too, but I'm still puzzling that out. I'll keep you posted.

@gofal 
Coming back to this, I spoke with the original developer that customized this code.
From his research, he found that VOILutSequence actually takes precedence over the other window level parameters if it is present. (You still need the Modality LUT though) 
That is why our GrayScaleRenderOptions.FromDataset short circuits if this sequence is present. Unfortunately I have yet to find the exact phrasing in the standard where this is specified, and he did not find it immediately again either.
I also dived into the DCMTK code where the rendering happens, and from my limited C++ knowledge, it does look like they do something similar. See this code specifically: https://github.com/DCMTK/dcmtk/blob/6e47fa05340431dadc175603ce8fdbfbfcb0b25c/dcmjpeg/libsrc/djcodece.cc#LL914C32-L914C32 

As for this pull request, I think you might have the same mistake I always made. :-) 
It might be better to chop this pull request up into smaller parts, because some parts are obviously very good (DicomImage caching) and other parts are still up for debate (the extension methods). By splitting it up, we can make intermediate progress on the things we agree on.

I am willing to help of course, since you already did a monstrous amount of effort here. :-)

mrbean-bremen commented:
> From his research, he found that VOILutSequence actually takes precedence over the other window level parameters if it is present.

Interesting. We alwaays have treated both as equal (e.g. any may be applied), and which one is used as the initial value was a matter of configuration.

The standard just says:
> If any VOI LUT Table is included by an Image, a Window Width and Window Center or the VOI LUT Table, but not both, may be applied to the Image for display. Inclusion of both indicates that multiple alternative views may be presented.

which doesn't say anything about which one shall be used as the primary value, but maybe there is someting elsewhere...


> I also dived into the DCMTK code where the rendering happens, and from my limited C++ knowledge, it does look like they do something similar. 

For me it looks like it depends on `windowType`, which is configurable.

I think in the end, it is the decision of the framework which VOI LUT to use if there is more than one, ideally giving the users the ability to choose for themselves. 

just my 2 cents...

gofal commented:
> > From his research, he found that VOILutSequence actually takes precedence over the other window level parameters if it is present.
> 
> Interesting. We alwaays have treated both as equal (e.g. any may be applied), and which one is used as the initial value was a matter of configuration.

Yes, both have to be supported. And if a VOILutSequence is present, then this will be applied by default.

> The standard just says:
> 
> > If any VOI LUT Table is included by an Image, a Window Width and Window Center or the VOI LUT Table, but not both, may be applied to the Image for display. Inclusion of both indicates that multiple alternative views may be presented.
> 
> which doesn't say anything about which one shall be used as the primary value, but maybe there is someting elsewhere...
> 
> > I also dived into the DCMTK code where the rendering happens, and from my limited C++ knowledge, it does look like they do something similar.
> 
> For me it looks like it depends on `windowType`, which is configurable.

an image can contain several windowing-values. the UI then shows a UI so that the user can select a windowing-setting. Similar it's with the VOI LUT. if there is a voilut and some windowings, then the user can select, which one to apply and of course can do manual adjustment.


In the current implementation the GrayscaleRenderOtpions.FromDataset is checking if there are windowwidth/windowcenter or smallestpixelvalue/largestpixelvalue or if the histrogram has to be evaluated. But on each of those conditions, the VoiLutSequence is always parsed. The GrayscalerenderOptions class has a property `UseVOILUT`. if there is a VoiLutSequence, then this property is set to `true`.
That means: the VOILUT is always parsed, and the parameters WindowWidth and WindowCenter are also always parsed. if the VoiLut is present, then is is initially applied, but it can be switched by the user.

In your implementation you only read the VOILUT and if present stop parsing further data. So your implementation is just ignoring the WindowCenter/WindowWidth if there would be some present.

gofal commented:
> I'd be interested to hear your thoughts on this implementation. This has been running in production for a few years already.

This is a very explit implementation, where you have a lot of code to iterate through the functionSequences and through the macros. all this code is just for explicitly reading the properties for grayscale render options. But for EnhancedImages (and this PR is for correctly displaying enhanced images) you also have to read plenty of tags like the PixelSpacing, the ImageOrientation, ImagePosition, Slicethickness, AnatomicRegion, etc.

The functional groups are organized in macros, but within the macros then there are the same tags as you would expect in root-dataset from Non-Enhanced datasets. 
If you do explicit, then everytime you want to access a item, you have to check if the item is in root, and otherwise search in Shared and in PerFrame if there is a macro (you have to know the Macro in which the tag is stored) and then within this macro you can check the item.
My approach was: first collect all items that are related to the frame by merging all macros, and then you have a list of tags. In our application we cache that list. and then we have the extension methods to get the value either from the root dataset or from the list of items stored in functional groups.

So you do not have convenience functions to access the functional items? Lets say, you have a dataset of a EnhancedMR and need to get just a value like 0008,1155 Referenced SOP Instance UID. 
With this extensionmethod here you can call `dataset.FunctionalGroupValues(0).GetSingleValue<DicomUID>(DicomTag.ReferencedSOPInstanceUID)` without iterating through various subclasses. What would the API in your implementation look like, to access such dicomitems contained in functional groups?

This is, why the convenience functions for accessing the functional group values is important in this PR to make EnhancedMR, EnhancedCT and BTO images render correctly. Because what makes those simple multiframe images so special, is, that many data is stored in those functional groups instead of in the root dataset. And furthermore, there can be different rendering options per frame in those functional groups. This is why this PR has those two changes. But they are very related to each other.

gofal committed:
implement AutoApplyLutToAllFrames

gofal committed:
undo usage of IsOneOf extension method

gofal committed:
- rename DicomDatasetGetterExtensions to DicomItemsCollectionExtensions
- add original getter methods back to DicomDataset. this duplicates code but preserves little bit better performance

gofal committed:
address codereview comments: rename variable and use array-methods instead of linq methods

gofal commented:
I did some benchmarking of all the benchmarks available in this project (dicomserver, parsing, rendering, open/save), one with the get-methods from DicomDataset and one with the new extension methods. the performance was approx 2-5% slower. I have not done any performance optimization in the extension method so far, but for now I reverted this change. Now all the get-methods are duplicated, one time directly in DicomDataset, and another time in extension method for IEnumerable<DicomItem>.
I will search for an other way to avoid this code duplication.

gofal committed:
use the functionalgroup values in FrameGeometry, to correctly calculate geometry in EnhancedCT or EnhancedMR

gofal committed:
fix imaging.desktop project settings after rebasing

gofal committed:
Siemens MAGNETOM Altea produce functional group modules with empty sequences

gofal committed:
images with ModalityLut but without windowing-data were rendered wrong. the Min- and Max values have to mapped in ModalityLut

gofal committed:
fix: there are images where windowing-data is in root dataset, but rescaleslope/rescaleintercept are in functional groups

tboby commented:
The two images I tried were from the JPEG2000 dataset (ftp://medical.nema.org/MEDICAL/Dicom/DataSets/WG04) hosted by nema but linked to from here https://www.dclunie.com/ 

[nemaj2kir.zip](https://github.com/fo-dicom/fo-dicom/files/12725974/nemaj2kir.zip)

I had to run it in a "WithTrancoder" test collection to get the efferent codecs, but then it fails with
```
FellowOakDicom.Imaging.DicomImagingException
Unsupported pipeline photometric interpretation: YBR Reversible Color Transformation (JPEG 2000)
   at FellowOakDicom.Imaging.DicomImage.CreatePipelineData(Int32 frame) in C:\Users\ThomasBoby\Git\fo-dicom\FO-DICOM.Core\Imaging\DicomImage.cs:line 470
```

gofal committed:
fix merge problem in changelog

gofal committed:
DicomDataset.FunctaionalGroupValues(frame) now returns a DicomDataset, so that the get-methods do not have to be duplicated.

gofal committed:
update changelog

gofal committed:
remove/undo commented code

gofal committed:
add "#nullable disable" to all new files
